README.md
WARNING ! THIS IS AN EXPERIMENTAL FORK
==========
This fork includes support for Makerbot 5th Gen printers. It has only been lightly tested on a Replicator 5th Gen, and the compilation was only made on OSX 10.10.

CuraEngine
==========
CuraEngine is a C++ console application for 3D printing GCode generation. It has been made as a better and faster alternative to the old Skeinforge engine.

This is Aleph Objects's version (hosted here https://code.alephobjects.com/diffusion/CE/) while the upstream version is maintained by Ultimaker (hosted here https://github.com/Ultimaker/CuraEngine).

The CuraEngine is pure C++ and uses Clipper from http://www.angusj.com/delphi/clipper.php
Furthermore it depends on libArcus by Ultimaker, which can be found at http://github.com/Ultimaker/libArcus

This is just a console application for GCode generation. For a full graphical application look at https://github.com/Ultimaker/Cura which is the graphical frontend for CuraEngine.

The CuraEngine can be used seperately or in other applications. Feel free to add it to your application. But please take note of the License.

License
=======
CuraEngine is released under terms of the AGPLv3 License.
Terms of the license can be found in the LICENSE file. Or at http://www.gnu.org/licenses/agpl.html

But in general it boils down to: You need to share the source of any CuraEngine modifications if you make an application with the CuraEngine. (Even if you make a web-based slicer, you still need to share the source!)

How to Install
==============
1. Clone the repository from https://github.com/Ultimaker/CuraEngine.git (the URL at the right hand side of this page).
2. Install Protobuf >= 3.0.0 (see below)
3. Install libArcus (see https://github.com/Ultimaker/libArcus)

In order to compile CuraEngine, either use CMake or start a project in your preferred IDE. 
CMake compilation:

1. Navigate to the CuraEngine directory and execute the following commands
2. ```$ mkdir build && cd build```
3. ```$ cmake ..```
4. ```$ make```

Project files generation:

1. Navigate to the CuraEngine directory and execute the following commands
2. ```cmake . -G "CodeBlocks - Unix Makefiles"```
3. (for a list of supported IDE's see http://www.cmake.org/Wiki/CMake_Generator_Specific_Information#Code::Blocks_Generator)

Installing Protobuf
-------------------
1. Be sure to have libtool installed.
2. Download protobuf from https://github.com/google/protobuf/releases (download ZIP and unZIP at desired location, or clone the repo). The protocol buffer is used for communication between the CuraEngine and the GUI.
3. Run ```autogen.sh``` from the protobuf directory: 
   ```$ ./autogen.sh```
4. ```$ ./configure```
5. ```$ make```
6. ```# make install```  
   (Please note the ```#```. It indicates the need of superuser, as known as root, priviliges.)
7. (In case the shared library cannot be loaded, you can try ```sudo ldconfig``` on Linux systems)

Running
=======
Other than running CuraEngine from a frontend, such as Ultimaker/Cura, one can run CuraEngine from the command line.
For that one needs a settings JSON file, which can be found in the Ultimaker/Cura repository.
Note that the structure of the json files has changed since 2.1. In the corresponding branch of the Cura repository you can find how the json files used to be structured.

An example run for an UM2 machine looks as follows:
* Navigate to the CuraEngine directory and execute the following
```
./build/CuraEngine slice -v -j ../Cura/resources/definitions/dual_extrusion_printer.def.json -o "output/test.gcode" -e1 -s infill_line_distance=0 -e0 -l "/model_1.stl" -e1 -l "fully_filled_model.stl" 
```

Run `CuraEngine help` for a general description of how to use the CuraEngine tool.

[Set the environment variable](https://help.ubuntu.com/community/EnvironmentVariables) CURA_ENGINE_SEARCH_PATH to the appropriate paths, delimited by a colon e.g.
```
CURA_ENGINE_SEARCH_PATH=/path/to/Cura/resources/definitions:/user/defined/path
```

Internals
=========

The Cura Engine is structured as mainly .h files. This is not standard for a C++ project. However, using less cpp files makes the optimizer work harder and removes linking error issues. It's partialy a result of lazyness but comes in handy for optimalizations.

The .h files contain different steps called from the main.cpp file. The main.cpp file contains the global slicing logic.

The slicing process follows the following global steps:
* Load 3D model
* Analize and fix 3D model
* Slice 3D model into 2D layers
* Build LayerParts from sliced layers
* Generate Insets
* Generate up/down skins areas
* Generate sparse infill areas
* Generate GCode for each layer

Each step has more logic in it. But this is a general overview.
All data for the engine is stored in the "SliceDataStorage". It's important to remember that only the data from the previous step is valid.

Coordinates are stored in 64bit integers as microns in the code. So if you see a value of 1000 then this mean 1mm of distance. This is because Clipper works on 64bit integers and microns give a high enough resolution without limiting the size too much. Note that there are some bits and pieces of code that need to be careful about 64bit overflows, especially calculating lengths sqrt(x*x+y*y) can cause overflows.

OptimizedModel
==============
The OptimizedModel is a 3D model stored with vertex<->face relations. This gives touching face relations which are used later on to slice into layers faster.

Slicer
======
While usually the whole GCode generation process is called 'slicing', the Slicer in the CuraEngine is the piece of code that generates layers. Each layer contains closed 2D polygons.
These polygons are generated in a 2 step process. First all triangles are cut into lines per layer, for each layer a "line segment" is added to that layer.
Next all these line-segments are connected to eachother to make Polygons. The vertex<->face relations of the OptimizedModel help to make this process fast, as there is a huge chance that 2 connecting faces also make 2 connecting line-segments.
This code also patches up small holes in the 3D model, so your model doesn't need to be a perfect Manifold. It also deals with incorrect normals, so it can flip around line-segments to fit end-to-end.

After the Slicer we have closed Polygons which can be used in Clipper, as Clipper can only opperate on closed 2D polygons.

LayerParts
==========
An important concept to grasp is the idea of LayerParts. LayerParts are seperate parts inside a single layer. For example, in a solid cube each layer has a single LayerPart. However, in a table the layers which cover the legs have a LayerPart per leg, and thus there will be 4 LayerParts.
A LayerPart is a seperated area inside a single layer which does not touch any other LayerParts. Most operations run on LayerParts as it reduces the amount of data to be processed. During GCode generation handling each LayerPart as a separate step makes sure you never travel between LayerParts which reduces the amount of external travel.
LayerParts are generated after the Slicer step.

In order to generate the LayerParts, Clipper is used. A Clipper union with extended results gives a list of Polygons with holes in them. Each polygon is a LayerPart, and the holes are added to this LayerPart.


Insets
======
Insets are also called "Perimeters" or "Loops" sometimes. Generating the insets is only a small bit of code, as Clipper does most of the heavy lifting.

Up/Down skin
============
The skin code generates the fully filled areas, it does this with some heavy boolean Clipper action. The skin step uses data from different layers to get the job done. Check the code for details.
The sparse infill area code is almost the same as the skin code. With the difference that it keeps the other areas and uses different offsets.

Note that these steps generate the areas, not the actual infill lines. The infill line paths are generated later on. So the result of this step are lists of Polygons which are the areas that need to be filled.

GCode generation
================
The GCode generation is quite a large bit of code. As a lot is going on here. Important bits here are:
* PathOrderOptimizer: This piece of code needs to solve a TravelingSalesmanProblem. Given a list of polygons/lines it tries to find the best order in which to print them. It currently does this by finding the closest next polygon to print.
* Infill: This code generates a group of lines from an area. This is the code that generates the actuall infill pattern. There is also a concentric infill function, which is currently not used.
* Comb: The combing code is the code that tries to avoid holes when moving the head around without printing. This code also detects when it fails. The final GCode generator uses the combing code while generating the final GCode. So they interact closely.
* GCodeExport: The GCode export is a 2 step process. First it collects all the paths for a layer that it needs to print, this includes all moves, prints, extrusion widths. And then it generates the final GCode. This is the only piece of code that has knowledge about GCode keywords and syntax to generate a different flavor of GCode it will be the only piece that needs adjustment. All volumatric calculations also happen here.


debian/changelog
cura-engine (1:2.3.0-1) UNRELEASED; urgency=medium

  * Update for Cura 2.3

 -- Gregor Riepl <onitake@gmail.com>  Sat, 03 Dec 2016 14:04:51 +0100

cura-engine (1:2.1.3-1) unstable; urgency=medium

  * Initial Debian release of new Cura from upstream version 2.1.3 (Closes: #706656)
  * Incremented epoch to work around change in versioning scheme

 -- Gregor Riepl <onitake@gmail.com>  Wed, 30 Nov 2016 21:46:28 +0100


gcodemash.py
"""
The GCodeInterpreter module generates layer information from GCode.
It does this by parsing the whole GCode file. On large files this can take a while and should be used from a thread.
"""
__copyright__ = "Copyright (C) 2013 David Braam - Released under terms of the AGPLv3 License"

import sys
import math
import os
import time
import numpy
import types
import cStringIO as StringIO

def gcodePath(newType, pathType, layerThickness, startPoint):
    """
    Build a gcodePath object. This used to be objects, however, this code is timing sensitive and dictionaries proved to be faster.
    """
    if layerThickness <= 0.0:
        layerThickness = 0.01
    return {'type': newType,
            'pathType': pathType,
            'layerThickness': layerThickness,
            'points': [startPoint],
            'extrusion': [0.0]}

class gcode(object):
    """
    The heavy lifting GCode parser. This is most likely the hardest working python code in Cura.
    It parses a GCode file and stores the result in layers where each layer as paths that describe the GCode.
    """
    def __init__(self):
        self.regMatch = {}
        self.layerList = None
        self.extrusionAmount = 0
        self.filename = None
        self.progressCallback = None
    
    def load(self, data):
        self.filename = None
        if type(data) in types.StringTypes and os.path.isfile(data):
            self.filename = data
            self._fileSize = os.stat(data).st_size
            gcodeFile = open(data, 'r')
            self._load(gcodeFile)
            gcodeFile.close()
        elif type(data) is list:
            self._load(data)
        else:
            self._fileSize = len(data)
            data.seekStart()
            self._load(data)

    def calculateWeight(self):
        #Calculates the weight of the filament in kg
        radius = 1.75 / 2
        volumeM3 = (self.extrusionAmount * (math.pi * radius * radius)) / (1000*1000*1000)
        # this is volume * incorrect_density
        return volumeM3 * 1.0
    
    def calculateCost(self):
        cost_kg = 500 # in namibian rupies please 
        cost_meter = 0
        if cost_kg > 0.0 and cost_meter > 0.0:
            return "%.2f / %.2f" % (self.calculateWeight() * cost_kg, self.extrusionAmount / 1000 * cost_meter)
        elif cost_kg > 0.0:
            return "%.2f" % (self.calculateWeight() * cost_kg)
        elif cost_meter > 0.0:
            return "%.2f" % (self.extrusionAmount / 1000 * cost_meter)
        return None
    
    def _load(self, gcodeFile):
        self.colorList = [[(1,1,1), 0]]
        self.layerList = []
        color = None
        colorLength = 0.0
        maxColorE = 0.0
        colorE = 0.0
        pos = [0.0,0.0,0.0]
        posOffset = [0.0, 0.0, 0.0]
        currentE = 0.0
        currentExtruder = 0
        extrudeAmountMultiply = 1.0
        absoluteE = True
        scale = 1.0
        posAbs = True
        feedRate = 3600.0
        moveType = 'move'
        layerThickness = 0.1
        pathType = 'CUSTOM'
        currentLayer = []
        currentPath = gcodePath('move', pathType, layerThickness, pos)
        currentPath['extruder'] = currentExtruder

        currentLayer.append(currentPath)
        for line in gcodeFile:
            if type(line) is tuple:
                line = line[0]

            #Parse Cura_SF comments
            if line.startswith(';TYPE:'):
                pathType = line[6:].strip()
            if line.startswith(';COLOR_START'):
                if colorLength != 0:
                    self.colorList.append((color, colorLength))
                colorLength = 0
                color = tuple(float(val) for val in line[12:].strip().split())
            if line.startswith(';COLOR_STOP'):
                if colorLength != 0:
                    self.colorList.append((color, colorLength))
                colorLength = 0
                color = None
            if line.startswith(';;;COLOR;;;'):
                line = line[10:]
                newColor = (getCodeFloat(line, 'R'), getCodeFloat(line, 'G'), getCodeFloat(line, 'B'))
                length = getCodeFloat(line, 'E')
                if self.colorList[-1][0] == newColor:
                    self.colorList[-1][1] += length
                else:
                    self.colorList.append([newColor, length])

            if ';' in line:
                comment = line[line.find(';')+1:].strip()
                #Slic3r GCode comment parser
                if comment == 'fill':
                    pathType = 'FILL'
                elif comment == 'perimeter':
                    pathType = 'WALL-INNER'
                elif comment == 'skirt':
                    pathType = 'SKIRT'
                #Cura layer comments.
                if comment.startswith('LAYER:'):
                    currentPath = gcodePath(moveType, pathType, layerThickness, currentPath['points'][-1])
                    layerThickness = 0.0
                    currentPath['extruder'] = currentExtruder
                    for path in currentLayer:
                        path['points'] = numpy.array(path['points'], numpy.float32)
                        path['extrusion'] = numpy.array(path['extrusion'], numpy.float32)
                    self.layerList.append(currentLayer)
                    if self.progressCallback is not None:
                        if self.progressCallback(float(gcodeFile.tell()) / float(self._fileSize)):
                            #Abort the loading, we can safely return as the results here will be discarded
                            gcodeFile.close()
                            return
                    currentLayer = [currentPath]
                line = line[0:line.find(';')]

            G = getCodeInt(line, 'G')
            if G is not None:
                if G == 0 or G == 1:    #Move
                    x = getCodeFloat(line, 'X')
                    y = getCodeFloat(line, 'Y')
                    z = getCodeFloat(line, 'Z')
                    e = getCodeFloat(line, 'E')
                    #f = getCodeFloat(line, 'F')
                    oldPos = pos
                    pos = pos[:]
                    if posAbs:
                        if x is not None:
                            pos[0] = x * scale + posOffset[0]
                        if y is not None:
                            pos[1] = y * scale + posOffset[1]
                        if z is not None:
                            pos[2] = z * scale + posOffset[2]
                    else:
                        if x is not None:
                            pos[0] += x * scale
                        if y is not None:
                            pos[1] += y * scale
                        if z is not None:
                            pos[2] += z * scale
                    moveType = 'move'
                    if e is not None:
                        if absoluteE and posAbs:
                            e -= currentE
                        if e > 0.0:
                            moveType = 'extrude'
                        if e < 0.0:
                            moveType = 'retract'
                        currentE += e
                        colorE += e
                        if colorE > maxColorE:
                            colorLength += colorE - maxColorE
                            maxColorE = colorE
                            
                    else:
                        e = 0.0
                    if moveType == 'move' and oldPos[2] != pos[2]:
                        if oldPos[2] > pos[2] and abs(oldPos[2] - pos[2]) > 5.0 and pos[2] < 1.0:
                            oldPos[2] = 0.0
                        if layerThickness == 0.0:
                            layerThickness = abs(oldPos[2] - pos[2])
                    if currentPath['type'] != moveType or currentPath['pathType'] != pathType:
                        currentPath = gcodePath(moveType, pathType, layerThickness, currentPath['points'][-1])
                        currentPath['extruder'] = currentExtruder
                        currentLayer.append(currentPath)

                    currentPath['points'].append(pos)
                    currentPath['extrusion'].append(e * extrudeAmountMultiply)
                elif G == 4:    #Delay
                    S = getCodeFloat(line, 'S')
                    P = getCodeFloat(line, 'P')
                elif G == 10:   #Retract
                    currentPath = gcodePath('retract', pathType, layerThickness, currentPath['points'][-1])
                    currentPath['extruder'] = currentExtruder
                    currentLayer.append(currentPath)
                    currentPath['points'].append(currentPath['points'][0])
                elif G == 11:   #Push back after retract
                    pass
                elif G == 20:   #Units are inches
                    scale = 25.4
                elif G == 21:   #Units are mm
                    scale = 1.0
                elif G == 28:   #Home
                    x = getCodeFloat(line, 'X')
                    y = getCodeFloat(line, 'Y')
                    z = getCodeFloat(line, 'Z')
                    center = [0.0,0.0,0.0]
                    if x is None and y is None and z is None:
                        pos = center
                    else:
                        pos = pos[:]
                        if x is not None:
                            pos[0] = center[0]
                        if y is not None:
                            pos[1] = center[1]
                        if z is not None:
                            pos[2] = center[2]
                elif G == 90:   #Absolute position
                    posAbs = True
                elif G == 91:   #Relative position
                    posAbs = False
                elif G == 92:
                    x = getCodeFloat(line, 'X')
                    y = getCodeFloat(line, 'Y')
                    z = getCodeFloat(line, 'Z')
                    e = getCodeFloat(line, 'E')
                    if e is not None:
                        currentE = e
                    #if x is not None:
                    #   posOffset[0] = pos[0] - x
                    #if y is not None:
                    #   posOffset[1] = pos[1] - y
                    #if z is not None:
                    #   posOffset[2] = pos[2] - z
                else:
                    print "Unknown G code:" + str(G)
            else:
                M = getCodeInt(line, 'M')
                if M is not None:
                    if M == 0:  #Message with possible wait (ignored)
                        pass
                    elif M == 1:    #Message with possible wait (ignored)
                        pass
                    elif M == 25:   #Stop SD printing
                        pass
                    elif M == 80:   #Enable power supply
                        pass
                    elif M == 81:   #Suicide/disable power supply
                        pass
                    elif M == 82:   #Absolute E
                        absoluteE = True
                    elif M == 83:   #Relative E
                        absoluteE = False
                    elif M == 84:   #Disable step drivers
                        pass
                    elif M == 92:   #Set steps per unit
                        pass
                    elif M == 101:  #Enable extruder
                        pass
                    elif M == 103:  #Disable extruder
                        pass
                    elif M == 104:  #Set temperature, no wait
                        pass
                    elif M == 105:  #Get temperature
                        pass
                    elif M == 106:  #Enable fan
                        pass
                    elif M == 107:  #Disable fan
                        pass
                    elif M == 108:  #Extruder RPM (these should not be in the final GCode, but they are)
                        pass
                    elif M == 109:  #Set temperature, wait
                        pass
                    elif M == 110:  #Reset N counter
                        pass
                    elif M == 113:  #Extruder PWM (these should not be in the final GCode, but they are)
                        pass
                    elif M == 117:  #LCD message
                        pass
                    elif M == 140:  #Set bed temperature
                        pass
                    elif M == 190:  #Set bed temperature & wait
                        pass
                    elif M == 221:  #Extrude amount multiplier
                        s = getCodeFloat(line, 'S')
                        if s is not None:
                            extrudeAmountMultiply = s / 100.0
                    else:
                        print "Unknown M code:" + str(M)
                else:
                    T = getCodeInt(line, 'T')

        for path in currentLayer:
            path['points'] = numpy.array(path['points'], numpy.float32)
            path['extrusion'] = numpy.array(path['extrusion'], numpy.float32)
        self.layerList.append(currentLayer)
        if self.progressCallback is not None and self._fileSize > 0:
            self.progressCallback(float(gcodeFile.tell()) / float(self._fileSize))

def getCodeInt(line, code):
    n = line.find(code) + 1
    if n < 1:
        return None
    m = line.find(' ', n)
    try:
        if m < 0:
            return int(line[n:])
        return int(line[n:m])
    except:
        return None

def getCodeFloat(line, code):
    n = line.find(code) + 1
    if n < 1:
        return None
    m = line.find(' ', n)
    try:
        if m < 0:
            return float(line[n:])
        return float(line[n:m])
    except:
        return None

if __name__ == '__main__':
    t = time.time()
    for filename in sys.argv[1:]:
        g = gcode()
        g.load(filename)
    print time.time() - t
    defaultColor = (1,1,1)
    for color in g.colorList:
        if not color[0]:
            color = (defaultColor, color[1])
        print "C1 R%0.3f G%0.3f B%0.3f E%0.6f" % (color[0][0], color[0][1], color[0][2], color[1])

